README.rst
==================================
A lazy connection context manager
==================================
A lazy connection context manager


使用说明
=======================
`LazyConnection` 是一个线程安全的连接管理器，它将会创建一个连接上下文(context)，这个上下文管理一堆在它上面注册的工场函数(factory)。

- 在context 的生命周期范围内，factory 将会根据使用到的情况按需创建(最多只会被创建一次), 通过以 context 上注册的名字作为属性来访问,
  可得到该 factory 的一个实例。

- 在context 的生命周期结束时，将会自动执行清理工作，factory 的实例会被销毁，如果有 close 方法也将被调用。

使用范例::

    from lazyconn import LazyConnection

    def create_test1_client():
        return snow(host='127.0.0.1', port=1234)

    def create_test2_client1():
        return wish(host='127.0.0.1', port=1235)

    # register
    LazyConnection.register_factory('test', create_test1_client)
    LazyConnection.register_factory('test2', create_test2_client)

    # context life
    with LazyConnection() as conn:
        conn.test1.hello('wayhome')

`lg` 是一个线程安全的全局对象，它只在连接上下文中有效，它的 **conn** 属性对应一个连接上下文。

.. note:: 在 context 生命周期外，不能使用此对象。

使用范例::

   from lazyconn.globals import lg

    with LazyConnection() as conn:
        conn.test1.hello('wayhome')
        lg.conn.test1.hello('wayhome2')


`thread_safe` 是一个装饰器, 被装饰的函数将会在一个连接上下文中执行。`lg` 对象可以在被装饰的函数内部使用。

使用范例::

    from lazyconn import thread_safe

    @thread_safe
    def test():
        lg.conn.test1.hello('wayhome3')


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2014-01-16)
++++++++++++++++++

* First release on PyPI.


tests/test_lazyconn.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_lazyconn
----------------------------------

Tests for `lazyconn` module.
"""

import unittest

from lazyconn import LazyConnection
from lazyconn import thread_safe
from lazyconn.globals import lg


class TService(object):

    def hello(self, name):
        return "hello {0}".format(name)


def create_test_client():
    return TService()


class TestLazyconn(unittest.TestCase):

    def setUp(self):
        LazyConnection.register_factory('test', create_test_client)

    def test_something(self):
        with LazyConnection() as conn:
            print conn.test.hello('wayhome')
            print lg.conn.test.hello('wayhome2')

    def test_decorator(self):
        @thread_safe
        def test():
            print lg.conn.test.hello("wayhome3")
        test()

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()


lazyconn/local.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
try:
    from greenlet import getcurrent as get_ident
except ImportError:  # pragma: no cover
    try:
        from thread import get_ident
    except ImportError:  # pragma: no cover
        from dummy_thread import get_ident


def try_import(module):
    from importlib import import_module
    try:
        return import_module(module)
    except ImportError:
        pass


def release_local(local):
    """Releases the contents of the local for the current context.
    This makes it possible to use locals without a manager.

    Example::

        >>> loc = Local()
        >>> loc.foo = 42
        >>> release_local(loc)
        >>> hasattr(loc, 'foo')
        False

    With this function one can release :class:`Local` objects as well
    as :class:`StackLocal` objects.  However it is not possible to
    release data held by proxies that way, one always has to retain
    a reference to the underlying local object in order to be able
    to release it.

    .. versionadded:: 0.6.1
    """
    local.__release_local__()


class Local(object):
    __slots__ = ('__storage__', '__ident_func__')

    def __init__(self):
        object.__setattr__(self, '__storage__', {})
        object.__setattr__(self, '__ident_func__', get_ident)

    def __iter__(self):
        return iter(self.__storage__.items())

    def __call__(self, proxy):
        """Create a proxy for a name."""
        return LocalProxy(self, proxy)

    def __release_local__(self):
        self.__storage__.pop(self.__ident_func__(), None)

    def __getattr__(self, name):
        try:
            return self.__storage__[self.__ident_func__()][name]
        except KeyError:
            raise AttributeError(name)

    def __setattr__(self, name, value):
        ident = self.__ident_func__()
        storage = self.__storage__
        try:
            storage[ident][name] = value
        except KeyError:
            storage[ident] = {name: value}

    def __delattr__(self, name):
        try:
            del self.__storage__[self.__ident_func__()][name]
        except KeyError:
            raise AttributeError(name)


class LocalStack(object):

    """This class works similar to a :class:`Local` but keeps a stack
    of objects instead.  This is best explained with an example::

        >>> ls = LocalStack()
        >>> ls.push(42)
        >>> ls.top
        42
        >>> ls.push(23)
        >>> ls.top
        23
        >>> ls.pop()
        23
        >>> ls.top
        42

    They can be force released by using a :class:`LocalManager` or with
    the :func:`release_local` function but the correct way is to pop the
    item from the stack after using.  When the stack is empty it will
    no longer be bound to the current context (and as such released).

    By calling the stack without arguments it returns a proxy that resolves to
    the topmost item on the stack.

    .. versionadded:: 0.6.1
    """

    def __init__(self):
        self._local = Local()

    def __release_local__(self):
        self._local.__release_local__()

    def _get__ident_func__(self):
        return self._local.__ident_func__

    def _set__ident_func__(self, value):
        object.__setattr__(self._local, '__ident_func__', value)
    __ident_func__ = property(_get__ident_func__, _set__ident_func__)
    del _get__ident_func__, _set__ident_func__

    def __call__(self):
        def _lookup():
            rv = self.top
            if rv is None:
                raise RuntimeError('object unbound')
            return rv
        return LocalProxy(_lookup)

    def push(self, obj):
        """Pushes a new item to the stack"""
        rv = getattr(self._local, 'stack', None)
        if rv is None:
            self._local.stack = rv = []
        rv.append(obj)
        return rv

    def pop(self):
        """Removes the topmost item from the stack, will return the
        old value or `None` if the stack was already empty.
        """
        stack = getattr(self._local, 'stack', None)
        if stack is None:
            return None
        elif len(stack) == 1:
            release_local(self._local)
            return stack[-1]
        else:
            return stack.pop()

    @property
    def top(self):
        """The topmost item on the stack.  If the stack is empty,
        `None` is returned.
        """
        try:
            return self._local.stack[-1]
        except (AttributeError, IndexError):
            return None


class LocalProxy(object):

    """Code stolen from werkzeug.local.LocalProxy."""
    __slots__ = ('__local', '__dict__', '__name__')

    def __init__(self, local, name=None):
        object.__setattr__(self, '_LocalProxy__local', local)
        object.__setattr__(self, '__name__', name)

    def _get_current_object(self):
        """Return the current object.  This is useful if you want the real
        object behind the proxy at a time for performance reasons or because
        you want to pass the object into a different context.
        """
        if not hasattr(self.__local, '__release_local__'):
            return self.__local()
        try:
            return getattr(self.__local, self.__name__)
        except AttributeError:
            raise RuntimeError('no object bound to %s' % self.__name__)

    @property
    def __dict__(self):
        try:
            return self._get_current_object().__dict__
        except RuntimeError:
            raise AttributeError('__dict__')

    def __repr__(self):
        try:
            obj = self._get_current_object()
        except RuntimeError:
            return '<%s unbound>' % self.__class__.__name__
        return repr(obj)

    def __nonzero__(self):
        try:
            return bool(self._get_current_object())
        except RuntimeError:
            return False

    def __unicode__(self):
        try:
            return unicode(self._get_current_object())
        except RuntimeError:
            return repr(self)

    def __dir__(self):
        try:
            return dir(self._get_current_object())
        except RuntimeError:
            return []

    def __getattr__(self, name):
        if name == '__members__':
            return dir(self._get_current_object())
        return getattr(self._get_current_object(), name)

    def __setitem__(self, key, value):
        self._get_current_object()[key] = value

    def __delitem__(self, key):
        del self._get_current_object()[key]

    def __setslice__(self, i, j, seq):
        self._get_current_object()[i:j] = seq

    def __delslice__(self, i, j):
        del self._get_current_object()[i:j]

    __setattr__ = lambda x, n, v: setattr(x._get_current_object(), n, v)
    __delattr__ = lambda x, n: delattr(x._get_current_object(), n)
    __str__ = lambda x: str(x._get_current_object())
    __lt__ = lambda x, o: x._get_current_object() < o
    __le__ = lambda x, o: x._get_current_object() <= o
    __eq__ = lambda x, o: x._get_current_object() == o
    __ne__ = lambda x, o: x._get_current_object() != o
    __gt__ = lambda x, o: x._get_current_object() > o
    __ge__ = lambda x, o: x._get_current_object() >= o
    __cmp__ = lambda x, o: cmp(x._get_current_object(), o)
    __hash__ = lambda x: hash(x._get_current_object())
    __call__ = lambda x, *a, **kw: x._get_current_object()(*a, **kw)
    __len__ = lambda x: len(x._get_current_object())
    __getitem__ = lambda x, i: x._get_current_object()[i]
    __iter__ = lambda x: iter(x._get_current_object())
    __contains__ = lambda x, i: i in x._get_current_object()
    __getslice__ = lambda x, i, j: x._get_current_object()[i:j]
    __add__ = lambda x, o: x._get_current_object() + o
    __sub__ = lambda x, o: x._get_current_object() - o
    __mul__ = lambda x, o: x._get_current_object() * o
    __floordiv__ = lambda x, o: x._get_current_object() // o
    __mod__ = lambda x, o: x._get_current_object() % o
    __divmod__ = lambda x, o: x._get_current_object().__divmod__(o)
    __pow__ = lambda x, o: x._get_current_object() ** o
    __lshift__ = lambda x, o: x._get_current_object() << o
    __rshift__ = lambda x, o: x._get_current_object() >> o
    __and__ = lambda x, o: x._get_current_object() & o
    __xor__ = lambda x, o: x._get_current_object() ^ o
    __or__ = lambda x, o: x._get_current_object() | o
    __div__ = lambda x, o: x._get_current_object().__div__(o)
    __truediv__ = lambda x, o: x._get_current_object().__truediv__(o)
    __neg__ = lambda x: -(x._get_current_object())
    __pos__ = lambda x: +(x._get_current_object())
    __abs__ = lambda x: abs(x._get_current_object())
    __invert__ = lambda x: ~(x._get_current_object())
    __complex__ = lambda x: complex(x._get_current_object())
    __int__ = lambda x: int(x._get_current_object())
    __long__ = lambda x: long(x._get_current_object())
    __float__ = lambda x: float(x._get_current_object())
    __oct__ = lambda x: oct(x._get_current_object())
    __hex__ = lambda x: hex(x._get_current_object())
    __index__ = lambda x: x._get_current_object().__index__()
    __coerce__ = lambda x, o: x.__coerce__(x, o)
    __enter__ = lambda x: x.__enter__()
    __exit__ = lambda x, *a, **kw: x.__exit__(*a, **kw)


lazyconn/lazyconn.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
LazyConnection是一个线程安全的连接管理器，它将会创建一个连接上下文(context)，这个上下文管理一堆在它上面注册的工场函数(factory)。

- 在context 的生命周期范围内，factory 将会根据使用到的情况按需创建(最多只会被创建一次), 通过以 context 上注册的名字作为属性来访问,
  可得到该 factory 的一个实例。

- 在context 的生命周期结束时，将会自动执行清理工作，factory 的实例会被销毁，如果有 close 方法也将被调用。

使用范例::

    from lazyconn import LazyConnection

    def create_test1_client():
        return snow(host='127.0.0.1', port=1234)

    def create_test2_client1():
        return wish(host='127.0.0.1', port=1235)

    # register
    LazyConnection.register_factory('test', create_test1_client)
    LazyConnection.register_factory('test2', create_test2_client)

    # context life
    with LazyConnection() as conn:
        conn.test1.hello('wayhome')
"""
import weakref
from .globals import _request_ctx_stack
from .local import Local


class Magic(object):

    def __init__(self, instance):
        self.instance = instance

    def __getattr__(self, name):
        if name not in self.instance.lg.connections:
            factory = self.instance._factories[name]
            self.instance.lg.connections[name] = factory()
        return self.instance.lg.connections[name]


class LazyConnection(object):
    _factories = {}

    def __init__(self):
        self.lg = Local()
        self.lg.connections = {}
        self.lg.conn = weakref.proxy(self)

    @classmethod
    def register_factory(cls, name, factory):
        cls._factories[name] = factory

    def __getattr__(self, name):
        if name in self._factories:
            return Magic(self).__getattr__(name)
        return super(LazyConnection, self).__getattr__(name)

    def __enter__(self):
        self.push()
        return self

    def push(self):
        """Binds the request context to the current context."""
        _request_ctx_stack.push(self)

    def pop(self):
        """Pops the request context and unbinds it by doing that.
        """
        _request_ctx_stack.pop()

    def __exit__(self, exc_type, exc_value, tb):
        for conn in self.lg.connections:
            if hasattr(conn, 'close'):
                conn.close()
        del self.lg.connections
        self.lg.__release_local__()


def thread_safe(func):
    def wrapper(*args, **kwargs):
        with LazyConnection():
            return func(*args, **kwargs)
    return wrapper


if __name__ == '__main__':
    class TService(object):

        def hello(self, name):
            return "hello {0}".format(name)

    def create_test_client():
        return TService()

    LazyConnection.register_factory('test', create_test_client)
    with LazyConnection() as conn:
        print conn.test.hello('wayhome')
